src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -54,35 +179,650 @@ private static void VerifyAnalyzersArgument(ImmutableArray<DiagnosticAnalyzer> a
+        }
+
+        /// <summary>
+        /// Returns semantic diagnostics produced by all <see cref="Analyzers"/> from analyzing the given <paramref name="model"/>, optionally scoped to a <paramref name="filterSpan"/>.
+        /// Depending on analyzers' behavior, returned diagnostics can have locations outside the tree,
+        /// and some diagnostics that would be reported for the tree by an analysis of the complete compilation
+        /// can be absent.
+        /// </summary>
+        /// <param name="model">Semantic model representing the syntax tree to analyze.</param>
+        /// <param name="filterSpan">An optional span within the tree to scope analysis.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        public async Task<ImmutableArray<Diagnostic>> GetAnalyzerSemanticDiagnosticsAsync(SemanticModel model, TextSpan? filterSpan, CancellationToken cancellationToken)
+        {
+            VerifyModel(model);
+
+            return await GetAnalyzerSemanticDiagnosticsCoreAsync(model, filterSpan, Analyzers, cancellationToken).ConfigureAwait(false);
         }
 
         /// <summary>
-        /// Returns diagnostics produced by diagnostic analyzers.
+        /// Returns semantic diagnostics produced by all <see cref="Analyzers"/> from analyzing the given <paramref name="model"/>, optionally scoped to a <paramref name="filterSpan"/>.
+        /// Depending on analyzers' behavior, returned diagnostics can have locations outside the tree,
+        /// and some diagnostics that would be reported for the tree by an analysis of the complete compilation
+        /// can be absent.
         /// </summary>
-        public async Task<ImmutableArray<Diagnostic>> GetAnalyzerDiagnosticsAsync()
+        /// <param name="model">Semantic model representing the syntax tree to analyze.</param>
+        /// <param name="filterSpan">An optional span within the tree to scope analysis.</param>
+        /// <param name="analyzers">Analyzers whose diagnostics are required. All the given analyzers must be from the analyzers passed into the constructor of <see cref="CompilationWithAnalyzers"/>.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        public async Task<ImmutableArray<Diagnostic>> GetAnalyzerSemanticDiagnosticsAsync(SemanticModel model, TextSpan? filterSpan, ImmutableArray<DiagnosticAnalyzer> analyzers, CancellationToken cancellationToken)
         {
+            VerifyModel(model);
+            VerifyExistingAnalyzersArgument(analyzers);
+
+            return await GetAnalyzerSemanticDiagnosticsCoreAsync(model, filterSpan, analyzers, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task<ImmutableArray<Diagnostic>> GetAnalyzerSemanticDiagnosticsCoreAsync(SemanticModel model, TextSpan? filterSpan, ImmutableArray<DiagnosticAnalyzer> analyzers, CancellationToken cancellationToken)
```

Comment:
@JohnHamby @srivatsn We actually don't need this API to take a `SemanticModel`, it could just be the underlying `SyntaxTree`. We need to operate on the mapped semantic model on the private compilation instance which has the event queue, so need to invoke `_compilation.GetSemanticModel(model.SyntaxTree).GetDiagnostics()` to force event generation for the tree.